packages/babel-parser/src/parser/expression.js

```diff
@@ -545,18 +549,25 @@ export default class ExpressionParser extends LValParser {
       return expr;
     }
 
-    return this.parseSubscripts(expr, startPos, startLoc);
+    return this.parseSubscripts(
+      expr,
+      startPos,
+      startLoc,
+      null,
+      asyncArrowFuncForbidden,
+    );
   }
 
   parseSubscripts(
     base: N.Expression,
     startPos: number,
     startLoc: Position,
     noCalls?: ?boolean,
+    asyncArrowFuncForbidden?: ?boolean,
   ): N.Expression {
     const state = {
       optionalChainMember: false,
-      maybeAsyncArrow: this.atPossibleAsync(base),
+      maybeAsyncArrow: this.atPossibleAsync(base) && !asyncArrowFuncForbidden,
```

Comment:
Can we simplify a bit to
```suggestion
      maybeAsyncArrow: this.atPossibleAsync(base) && startPos === this.state.potentialArrowAt,
```

Reply:
I find your suggestion also easier to understand :+1:

Reply:
I realize that `atPossibleAsync` is only used to detect whether there is an async arrow ahead. I suggest we rename `atPossibleAsync` to `atPossibleAsyncArrow`. And the condition `base.start === this.state.potentialArrowAt` can be added to current `atPossibleAsync`.

By doing so we can also fix another related case for free
```typescript
4 + async<number>() => 2 
// This typescript snippet should throw but now it is parsed successfully
```